syntax = "proto3";

package user;

service UserService {
  rpc GetUser (UserRequest) returns (UserResponse);
  rpc CreateUser (CreateUserRequest) returns (UserResponse);
  rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
  rpc CreateUserRating (CreateUserRatingRequest) returns (UserRatingResponse);
  rpc GetUserRatings (UserRequest) returns (UserRatingsResponse);
  rpc FollowUser (FollowUserRequest) returns (FollowUserResponse);
  rpc GetUserFollowers (UserRequest) returns (UserFollowersResponse);
  rpc GetUserFollowing (UserRequest) returns (UserFollowersResponse);
  rpc CheckFollowingStatus (CheckFollowingRequest) returns (FollowUserResponse);
}

message UserRequest {
  int64 id = 1;
}

message CreateUserRequest {
  string first_name = 1;
  string last_name = 2;
  string email = 3;
  string phone = 4;
  string password = 5;
  string role = 6;
  string address = 7;
  float latitude = 8;
  float longitude = 9;
  string bio = 10;
}

message DeleteUserRequest {
  int64 user_id = 1;
}

message DeleteUserResponse {
  bool success = 1;
  string message = 2;
}

message UserResponse {
  int64 id = 1;
  string first_name = 2;
  string last_name = 3;
  string email = 4;
  string phone = 5;
  string profile_photo = 6;
  string role = 7;
  string address = 8;
  float latitude = 9;
  float longitude = 10;
  string bio = 11;
  bool isActive = 12;
  bool email_verified = 13;
  bool phone_verified = 14;
  string created_at = 15;
}

message CreateUserRatingRequest {
  int64 rated_user_id = 1;
  int64 rated_by_user_id = 2;
  int32 rating_value = 3;
  string review = 4;
  string rating_type = 5;
}

message UserRatingResponse {
  int64 id = 1;
  int64 rated_user_id = 2;
  int64 rated_by_user_id = 3;
  int32 rating_value = 4;
  string review = 5;
  string rating_type = 6;
  string created_at = 7;
  string updated_at = 8;
}

message UserRatingsResponse {
  repeated UserRatingResponse ratings = 1;
}

message FollowUserRequest {
  int64 user_id = 1;
  int64 following_id = 2;
}

message FollowUserResponse {
  int64 id = 1;
  int64 user_id = 2;
  int64 following_id = 3;
  string status = 4;
  string followed_at = 5;
}

message UserFollowersResponse {
  repeated FollowUserResponse followers = 1;
}

message CheckFollowingRequest {
  int64 user_id = 1;
  int64 following_id = 2;
}